heat/core/linalg/svdtools.py

```diff
@@ -0,0 +1,500 @@
+
+from ..communication import MPICommunication
+from ..dndarray import DNDarray
+from .. import factories
+from .. import types
+from ..linalg import matmul, vector_norm
+from ..indexing import where
+from ..random import randn
+
+from ..manipulations import vstack, hstack, diag, balance
+
+from .. import statistics
+from math import log, ceil, floor, sqrt
+
+
+__all__ = ["hsvd_rank", "hsvd_rtol", "hsvd"]
+
+
+#########################################################################################
+# user-friendly versions of hSVD
+#########################################################################################
+
+
+def hsvd_rank(
+    A: DNDarray,
+    maxrank: int,
+    full: bool = False,
+    maxmergedim: Optional[int] = None,
+    safetyshift: int = 5,
+    silent: bool = True,
+) -> Union[
+    Tuple[DNDarray, DNDarray, DNDarray, float], Tuple[DNDarray, DNDarray, DNDarray], DNDarray
+]:
+    """
+    Hierarchical SVD (hSVD) with prescribed truncation rank maxrank.
+    If A = U diag(sigma) V^T is the true SVD of A, this routine computes an approximation for U[:,:maxrank] (and sigma[:maxrank], V[:,:maxrank])
+
+    The accuracy of this approximation depends on the structure of A ("low-rank" is best) and appropriate choice of parameters.
+
+    Parameters
+    ----------
+    A : DNDarray
+        2D-array (float32/64) of which the hSVD has to be computed.
+    maxrank : int
+        truncation rank.
+    full : bool, optional
+        full=True implies that also Sigma and V are computed and returned. The default is False.
+    maxmergedim : int, optional
+        maximal size of the concatenation matrices during the merging procedure. The default is None and results in an appropriate choice depending on the size of the local slices of A and maxrank.
+        Too small choices for this parameter will result in failure if the maximal size of the concatenation matrices does not allow to merge at least two matrices. Too large choices for this parameter can cause memory errors if the resulting merging problem becomes too large.
+    safetyshift : int, optional
+        Increases the actual truncation rank within the computations by a safety shift. The default is 5.
+    silent : bool, optional
+        silent=False implies that some information on the computations are printed. The default is True.
+
+    Returns
+    -------
+    (Union[    Tuple[DNDarray, DNDarray, DNDarray, float], Tuple[DNDarray, DNDarray, DNDarray], DNDarray])
+        if full=True: U, Sigma, V, a-posteriori error estimate for the reconstruction error ||A-U Sigma V^T ||_F / ||A||_F (computed according to [2] along the "true" merging tree).
+        if full=False: U, a-posteriori error estimate
+
+    Notes
+    -------
+    SVDs of matrices up to size A.size[0] x [2 * (maxrank + safetyshift) + 1] need to be computed on a single MPI-process during merging; too large values for maxrank and safetyshift therefore may result in memory issues.
```

Comment:
Is that `A.shape[0]` ?

Reply:
If I understand correctly, matrices with split 0 will be transposed so that it that case the size of the local matrix would be`(A.shape[1]  , [2 * (maxrank + safetyshift) + 1]).

Here I would just write that the size of each process-local SVD is proportional to 2 * (maxrank + safetyshift) so be conservative with those two parameters.

Reply:
Has been taken into account.